File: pilot_phase2_poc/partner-source/partner-source-springboot/src/main/java/com/waypoint/partnersource/shared/security/AccessPolicy.java
package com.waypoint.partnersource.shared.security;

import com.waypoint.partnersource.assignment.domain.AssignmentAuthorizationPolicy;
import com.waypoint.partnersource.assignment.repository.InMemoryAssignmentRepository;
import com.waypoint.partnersource.order.repository.InMemoryOrderRepository;
import org.springframework.stereotype.Component;

@Component
public class AccessPolicy {
    private final InMemoryOrderRepository orderRepository;
    private final InMemoryAssignmentRepository assignmentRepository;
    private final AssignmentAuthorizationPolicy assignmentAuthorizationPolicy;

    public AccessPolicy(
            InMemoryOrderRepository orderRepository,
            InMemoryAssignmentRepository assignmentRepository,
            AssignmentAuthorizationPolicy assignmentAuthorizationPolicy
    ) {
        this.orderRepository = orderRepository;
        this.assignmentRepository = assignmentRepository;
        this.assignmentAuthorizationPolicy = assignmentAuthorizationPolicy;
    }

    public boolean canReadDriverResource(AuthenticatedPrincipal principal, String driverId) {
        return principal.role() == ActorRole.DELIVERY_DRIVER
                && principal.actorType() == PrincipalActorType.DRIVER
                && principal.actorId().equals(driverId);
    }

    public boolean canReadOrder(AuthenticatedPrincipal principal, String orderId) {
        if (principal.role() == ActorRole.CUSTOMER_SERVICE_AGENT) {
            return true;
        }

        if (principal.role() != ActorRole.DELIVERY_DRIVER) {
            return false;
        }

        if (orderRepository.findById(orderId).isEmpty()) {
            return true;
        }

        return assignmentAuthorizationPolicy.canDriverUpdateOrder(
                principal.actorId(),
                orderId,
                assignmentRepository.findByOrderId(orderId)
        );
    }

    public boolean canCreateStatusEvent(AuthenticatedPrincipal principal) {
        return principal.role() == ActorRole.DELIVERY_DRIVER
                && principal.actorType() == PrincipalActorType.DRIVER;
    }

    public boolean canSubmitDriverId(AuthenticatedPrincipal principal, String driverId) {
        return principal.role() == ActorRole.DELIVERY_DRIVER
                && principal.actorType() == PrincipalActorType.DRIVER
                && principal.actorId().equals(driverId);
    }
}

File: pilot_phase2_poc/partner-source/partner-source-fastapi/app/security/access_policy.py
from app.domain.policies import AssignmentAuthorizationPolicy
from app.repositories.assignments import InMemoryAssignmentRepository
from app.repositories.orders import InMemoryOrderRepository
from app.security.principal import ActorRole, AuthenticatedPrincipal, PrincipalActorType


class AccessPolicy:
    def __init__(
        self,
        order_repository: InMemoryOrderRepository,
        assignment_repository: InMemoryAssignmentRepository,
    ) -> None:
        self._order_repository = order_repository
        self._assignment_repository = assignment_repository
        self._assignment_policy = AssignmentAuthorizationPolicy()

    def can_read_driver_resource(self, principal: AuthenticatedPrincipal, driver_id: str) -> bool:
        return (
            principal.role == ActorRole.DELIVERY_DRIVER
            and principal.actorType == PrincipalActorType.DRIVER
            and principal.actorId == driver_id
        )

    def can_read_order(self, principal: AuthenticatedPrincipal, order_id: str) -> bool:
        if principal.role == ActorRole.CUSTOMER_SERVICE_AGENT:
            return True
        if principal.role != ActorRole.DELIVERY_DRIVER:
            return False
        if self._order_repository.find_by_id(order_id) is None:
            return True
        return self._assignment_policy.can_driver_update_order(
            principal.actorId,
            order_id,
            self._assignment_repository.find_by_order_id(order_id),
        )

    def can_create_status_event(self, principal: AuthenticatedPrincipal) -> bool:
        return principal.role == ActorRole.DELIVERY_DRIVER and principal.actorType == PrincipalActorType.DRIVER

    def can_submit_driver_id(self, principal: AuthenticatedPrincipal, driver_id: str) -> bool:
        return (
            principal.role == ActorRole.DELIVERY_DRIVER
            and principal.actorType == PrincipalActorType.DRIVER
            and principal.actorId == driver_id
        )
